File: voyago_iOS/voyago_iOS/Models/VoyagoImage.swift
//
//  VoyagoImage.swift
//  voyago_iOS
//
//  Created by Yazan Ghunaim on 1/14/25.
//

import Foundation

/// Data connected with an image.
struct VoyagoImage: Codable {
    let id = UUID()  // id of image
    let username: String  // Photographer's username
    let unsplashProfile: String?  // Optional Unsplash profile URL

    private let urls: [String: String]  // [resolution: url]

    init(username: String, unsplashProfile: String? = nil, urls: [String: String]) {
        self.username = username
        self.unsplashProfile = unsplashProfile
        self.urls = urls
    }
}

extension VoyagoImage {
    // TODO: - Enum for sizes rather than literals
    var smallUrl: String { urls["small"] ?? "" }
    var regularUrl: String { urls["regular"] ?? "" }
    var fullUrl: String { urls["full"] ?? "" }
}

extension VoyagoImage {
    enum CodingKeys: String, CodingKey {
        case urls
        case username
        case unsplashProfile = "unsplash_profile"
    }
}

File: backend/schemas/voyago_image.py
"""pydantic models related to image data"""
from typing import Dict, Optional

from pydantic import BaseModel


class VoyagoImage(BaseModel):
    """Data connected with an image

    resolution: can be (regular, full, small)

    username | unsplash_profile:
        due to unsplash guidelines user must be credited
        in the future I hope voyago is run by users and not relying on unsplash
    """
    urls: Dict[str, str]  # [resolution : url]

    username: str  # photographer username
    unsplash_profile: Optional[str]
    # TODO: location needs separate call to getimage
    # location: str
